framework/bootstrap/configcontroller.go

```diff
@@ -0,0 +1,557 @@
+	"net/url"
+	"slime.io/slime/framework/bootstrap/viewstore"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/rest"
+	bootconfig "slime.io/slime/framework/apis/config/v1alpha1"
+	"slime.io/slime/framework/bootstrap/adsc"
+	"slime.io/slime/framework/bootstrap/collections"
+	"slime.io/slime/framework/bootstrap/resource"
+	"slime.io/slime/framework/bootstrap/serviceregistry/kube"
+	"slime.io/slime/framework/bootstrap/serviceregistry/model"
+	"slime.io/slime/framework/bootstrap/serviceregistry/serviceentry"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	mcpresource "istio.io/istio-mcp/pkg/config/schema/resource"
+	mcpc "istio.io/istio-mcp/pkg/mcp/client"
+	xdsc "istio.io/istio-mcp/pkg/mcp/xds/client"
+	mcpmodel "istio.io/istio-mcp/pkg/model"
+)
+
+const (
+	KubernetesConfigSourcePrefix = "k8s://"
+	McpOverXdsConfigSourcePrefix = "xds://"
+)
+
+type ConfigController interface {
+	viewstore.ViewerStore
+	RegisterEventHandler(kind resource.GroupVersionKind, handler func(resource.Config, resource.Config, Event))
+	Run(stop <-chan struct{})
+}
+
+type configController struct {
+	viewerStore        viewstore.ViewerStore
+	monitorControllers []*monitorController
+	stop               chan<- struct{}
+}
+
+func NewConfigController(ctx context.Context, config *bootconfig.Config, cfg *rest.Config) (ConfigController, error) {
+	var cc *configController
+	stopCh := make(chan struct{})
+	var mcs []*monitorController
+
+	configRevision := initConfigRevision(config.Global.IstioRev, config.Global.ConfigRev)
+	xdsSourceEnableIncPush := config.Global.Misc["xdsSourceEnableIncPush"] == "true"
+
+	// init all monitorControllers
+	imc, err := initIstioMonitorController(ctx)
+	if err != nil {
+		return nil, err
+	} else {
+		mcs = append(mcs, imc)
+	}
+
+	if len(config.Global.ConfigSources) > 0 {
+		for _, configSource := range config.Global.ConfigSources {
+			if strings.HasPrefix(configSource.Address, McpOverXdsConfigSourcePrefix) {
+				if mc, err := initXdsMonitorController(ctx, configSource); err != nil {
+					return nil, err
+				} else {
+					mcs = append(mcs, mc)
+				}
+				continue
+			}
+			if strings.HasPrefix(configSource.Address, KubernetesConfigSourcePrefix) {
+				if mc, err := initK8sMonitorController(ctx, configSource); err != nil {
+					return nil, err
+				} else {
+					mcs = append(mcs, mc)
+				}
+				continue
+			}
+		}
+	} else {
+		// init k8s for default
+		if mc, err := initK8sMonitorController(ctx, nil); err != nil {
+			return nil, err
+		} else {
+			mcs = append(mcs, mc)
+		}
+	}
+
+	vs, err := makeViewerStore(mcs)
+	if err != nil {
+		return nil, err
+	}
+
+	seLabelSelectorKeys := config.Global.Misc["seLabelSelectorKeys"]
+
+	cc = &configController{
+		viewerStore:        vs,
+		monitorControllers: mcs,
+		stop:               stopCh,
+	}
+
+	// use cc to register handler for istio resources
+	seToIstioResHandler := func(old resource.Config, cfg resource.Config, e Event) {
```

Comment:
actually this logic is not in the scope of `ConfigController`, we should split it into config part and service part like what istio does.

Reply:
add todo issue slime/slime/-/issues/137